telemetry/data_slow_query.go

```diff
@@ -0,0 +1,187 @@
+
+func (bucketMap SlowQueryBucket) String() string {
+	if bucketMap == nil {
+		return "nil"
+	}
+	var retStr string = "{"
+	for k, v := range bucketMap {
+		retStr += k + ":" + strconv.Itoa(v) + ","
+	}
+	retStr = retStr[:len(retStr)-1]
+	return retStr
+}
+
+const slowQueryBucketNum = 29 //prometheus.ExponentialBuckets(0.001, 2, 28), and 1 more +Inf
+
+var (
+	// LastSQBInfo records last statistic information of slow query buckets
+	LastSQBInfo SlowQueryBucket
+	// CurrentSQBInfo records current statitic information of slow query buckets
+	CurrentSQBInfo SlowQueryBucket
+)
+
+func getSlowQueryStats(ctx sessionctx.Context) (*slowQueryStats, error) {
+	slowQueryBucket, err := getSlowQueryBucket(ctx)
+	if err != nil {
+		logutil.BgLogger().Info(err.Error())
+		return nil, err
+	}
+
+	return &slowQueryStats{slowQueryBucket}, nil
+}
+
+// getSlowQueryBucket genenrates the delta SlowQueryBucket to report
+func getSlowQueryBucket(ctx sessionctx.Context) (*SlowQueryBucket, error) {
+	// update CurrentSQBInfo first, then gen delta
+	if err := updateCurrentSQB(ctx); err != nil {
+		return nil, err
+	}
+	delta := calculateDeltaSQB()
+	return delta, nil
+}
+
+// updateCurrentSQB records current slow query buckets
+func updateCurrentSQB(ctx sessionctx.Context) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			switch x := r.(type) {
+			case string:
+				err = errors.New(x)
+			case error:
+				err = x
+			default:
+				err = errors.New("unknown failure")
+			}
+		}
+	}()
+
+	value, err := querySlowQueryMetric(ctx) //TODO: judge error here
+	if err != nil && err != infosync.ErrPrometheusAddrIsNotSet {
+		logutil.BgLogger().Info("querySlowQueryMetric got error")
+		return err
+	}
+
+	if value.Type() != pmodel.ValVector {
+		return errors.New("Prom vector expected, got " + value.Type().String())
+	}
+	promVec := value.(pmodel.Vector)
+	for _, sample := range promVec {
+		metric := sample.Metric
+		bucketName := metric["le"] //hardcode bucket upper bound
+		CurrentSQBInfo[string(bucketName)] = int(sample.Value)
+	}
+	return nil
+}
+
+func querySlowQueryMetric(sctx sessionctx.Context) (result pmodel.Value, err error) {
+	// Add retry to avoid network error.
+	var prometheusAddr string
+	for i := 0; i < 5; i++ {
+		//TODO: the prometheus will be Integrated into the PD, then we need to query the prometheus in PD directly, which need change the quire API
+		prometheusAddr, err = infosync.GetPrometheusAddr()
+		if err == nil || err == infosync.ErrPrometheusAddrIsNotSet {
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+	if err != nil {
+		return nil, err
+	}
+	promClient, err := api.NewClient(api.Config{
+		Address: prometheusAddr,
+	})
+	if err != nil {
+		return nil, err
+	}
+	promQLAPI := promv1.NewAPI(promClient)
+	promQL := "tidb_server_slow_query_process_duration_seconds_bucket{sql_type=\"general\"}"
+
+	ts := time.Now()
+	// Add retry to avoid network error.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
```

Comment:
Maybe we should derive the context from sctx, so that when server is killed this can be killed as well.

BTW sometimes 1s timeout is not enough for a prometheus query. I suggest to set to at least 30s.

Reply:
There is no context in session.context, it's not derived from go routine context......